java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -103,13 +103,28 @@ class Guard extends ExprParent {
   }
 
   /**
-   * Gets the basic block containing this guard or the basic block containing
-   * the switch expression if the guard is a switch case.
+   * Gets the basic block containing this guard or the basic block that tests the
+   * applicability of this switch case -- for a pattern case this is the case statement
+   * itself; for a non-pattern case this is the most recent pattern case or the top of
+   * the switch block if there is none.
    */
   BasicBlock getBasicBlock() {
-    result = this.(Expr).getBasicBlock() or
-    result = this.(SwitchCase).getSwitch().getExpr().getBasicBlock() or
-    result = this.(SwitchCase).getSwitchExpr().getExpr().getBasicBlock()
+    // Not a switch case
+    result = this.(Expr).getBasicBlock()
+    or
+    // Return the closest pattern case statement before this one, including this one.
+    result =
+      max(int i, PatternCase c |
+        c = this.(SwitchCase).getSiblingCase(i) and i <= this.(SwitchCase).getCaseIndex()
```

Comment:
As noted at the definition of `getSiblingCase`, this is quadratic, and we'll want to implement this differently.